### PR TITLE
ISPN-12998 pass subject from task context

### PR DESCRIPTION
**Pull request by karesti**
https://issues.redhat.com/browse/ISPN-12998

### CODE REVIEW

**Comment by karesti** [comment]
Needs to be backported

File: tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java [hunk 1]
```diff
@@ -150,9 +150,12 @@ boolean containsScript(String taskName) {
          if (authorizationManager != null) {
             authorizationManager.checkPermission(AuthorizationPermission.EXEC, metadata.role().orElse(null));
          } else {
-            authorizer.checkPermission(AuthorizationPermission.EXEC, metadata.role().orElse(null));
+            if (context.getSubject().isPresent()) {
```
Comment (pruivo) [inline tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java:153]: this is the wrong fix. The problem is in the Rest enpoint.

Comment (karesti) [reply tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java:153]: but I'm not using REST but hotrod @pruivo

Comment (pruivo) [reply tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java:153]: just ignore me :rofl:

**Comment by pruivo** [comment]
it would be nice to have test/reproducer

**Comment by karesti** [comment]
@tristantarrant @pruivo reproducer added

**Comment by pruivo** [comment]
@karesti there is something wrong. If I roll back your fix, the "your" reproduce still passes.

**Comment by karesti** [comment]
@pruivo I tested without my change and I did see it failing with subject admin lacks admin permission. weird

**Comment by karesti** [comment]
@pruivo not the create script but the exec script

**Comment by karesti** [comment]
@pruivo create script is not failing I added for coverage

**Comment by karesti** [comment]
@pruivo nice catch, I removed the creation of the cache that was creating a secured cache and so entering the previous if instead of the one that does not concern not secured caches

**Review by pruivo — APPROVED** [review]

**Comment by karesti** [comment]
@tristantarrant how does the PR look?

File: server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java [hunk 2]
```diff
@@ -171,6 +171,52 @@ public void testRestWriterCannotReadImplicit() {
...
+
+   @Test
+   public void testExecScripts() {
+      InfinispanServerTestMethodRule serverTest = getServerTest();
+      RemoteCache cache = serverTest.hotrod().withClientConfiguration(hotRodBuilders.get(TestUser.ADMIN)).create();
+      String scriptName = serverTest.addScript(cache.getRemoteCacheManager(), "scripts/test.js");
+      Map params = new HashMap<>();
+      params.put("key", "k");
+      params.put("value", "v");
+      cache.getRemoteCacheManager().stop();
```
Comment (tristantarrant) [inline server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java:201]: No need to stop. RCMs created by the test methods are automatically stopped

File: server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java [hunk 3]
```diff
@@ -171,6 +171,52 @@ public void testRestWriterCannotReadImplicit() {
...
+         cacheExec.execute(scriptName, params);
+         cacheExec.getRemoteCacheManager().stop();
+      }
+
+      for (TestUser user : EnumSet.of(TestUser.MONITOR, TestUser.OBSERVER, TestUser.WRITER)) {
+         Exceptions.expectException(HotRodClientException.class, "(?s).*ISPN000287.*",
+                 () -> {
+                    RemoteCache cacheExec = serverTest.hotrod().withClientConfiguration(hotRodBuilders.get(user)).get();
+                    cacheExec.execute(scriptName, params);
+                    cacheExec.getRemoteCacheManager().stop();
```
Comment (tristantarrant) [inline server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java:214]: No need to stop. RCMs created by the test methods are automatically stopped

File: server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java [hunk 4]
```diff
@@ -171,6 +171,52 @@ public void testRestWriterCannotReadImplicit() {
...
+         serverTest.addScript(remoteCacheManager, "scripts/test.js");
+         remoteCacheManager.stop();
+      }
+
+      for (TestUser user : EnumSet.of(TestUser.MONITOR, TestUser.APPLICATION, TestUser.OBSERVER, TestUser.WRITER)) {
+         RemoteCacheManager remoteCacheManager = serverTest.hotrod().withClientConfiguration(hotRodBuilders.get(user)).createRemoteCacheManager();
+         Exceptions.expectException(HotRodClientException.class, "(?s).*ISPN000287.*",
+                 () -> serverTest.addScript(remoteCacheManager, "scripts/test.js")
+         );
+         remoteCacheManager.stop();
```
Comment (tristantarrant) [inline server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java:189]: No need to stop. RCMs created by the test methods are automatically stopped

File: server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java [hunk 5]
```diff
@@ -171,6 +171,52 @@ public void testRestWriterCannotReadImplicit() {
...
    }
 
+   @Test
+   public void testScriptUpload() {
+      InfinispanServerTestMethodRule serverTest = getServerTest();
+
+      for (TestUser user : EnumSet.of(TestUser.ADMIN, TestUser.DEPLOYER)) {
+         RemoteCacheManager remoteCacheManager = serverTest.hotrod().withClientConfiguration(hotRodBuilders.get(user)).createRemoteCacheManager();
+         serverTest.addScript(remoteCacheManager, "scripts/test.js");
+         remoteCacheManager.stop();
```
Comment (tristantarrant) [inline server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java:181]: No need to stop. RCMs created by the test methods are automatically stopped

**Comment by karesti** [comment]
I removed the stops

**Comment by karesti** [comment]
@tristantarrant is there anything else needed here ?

File: tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java [hunk 6]
```diff
@@ -148,11 +148,15 @@ boolean containsScript(String taskName) {
          AuthorizationManager authorizationManager = context.getCache().isPresent() ?
                SecurityActions.getAuthorizationManager(context.getCache().get().getAdvancedCache()) : null;
          if (authorizationManager != null) {
+            // when the cache is secured
             authorizationManager.checkPermission(AuthorizationPermission.EXEC, metadata.role().orElse(null));
          } else {
-            authorizer.checkPermission(AuthorizationPermission.EXEC, metadata.role().orElse(null));
+            if (context.getSubject().isPresent()) {
+               authorizer.checkPermission(context.getSubject().get(), AuthorizationPermission.EXEC);
```
Comment (danberindei) [inline tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingManagerImpl.java:155]: Katia, could you add a comment to explain why this is needed? 

Also, I noticed a similar pattern in `ServerTaskEngine.checkPermissions()`, maybe the same change is required there as well?

File: server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java [hunk 7]
```diff
@@ -171,6 +171,47 @@ public void testRestWriterCannotReadImplicit() {
...
+
+      for (TestUser user : EnumSet.of(TestUser.ADMIN, TestUser.APPLICATION, TestUser.DEPLOYER)) {
+         RemoteCache cacheExec = serverTest.hotrod().withClientConfiguration(hotRodBuilders.get(user)).get();
+         cacheExec.execute(scriptName, params);
+      }
+
+      for (TestUser user : EnumSet.of(TestUser.MONITOR, TestUser.OBSERVER, TestUser.WRITER)) {
+         Exceptions.expectException(HotRodClientException.class, "(?s).*ISPN000287.*",
+                 () -> {
+                    RemoteCache cacheExec = serverTest.hotrod().withClientConfiguration(hotRodBuilders.get(user)).get();
```
Comment (danberindei) [inline server/tests/src/test/java/org/infinispan/server/security/authorization/AbstractAuthorization.java:208]: This line should be outside `expectException()`, like in testScriptUpload

**Review by danberindei — CHANGES_REQUESTED** [review]

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**